task/control-plane-agent/src/mgs_sidecar.rs

```diff
@@ -166,19 +236,235 @@ impl MgsHandler {
+            return Err(GwMonorailError::InvalidVLAN);
+        };
+
+        if VLANS[vlan_index].trusted {
+            return Err(GwMonorailError::AlreadyTrusted);
+        }
+
+        let Some((last_challenge, challenge_time)) =
+            core::mem::take(&mut self.last_challenge)
+        else {
+            ringbuf_entry!(Trace::NoChallenge);
+            return Err(GwMonorailError::UnlockAuthFailed);
+        };
+
+        if challenge != last_challenge {
+            ringbuf_entry!(Trace::WrongChallenge);
+            return Err(GwMonorailError::UnlockAuthFailed);
+        }
+
+        let now = sys_get_timer().now;
+        if now >= challenge_time + CHALLENGE_EXPIRATION_TIME_SECS * 1000 {
+            ringbuf_entry!(Trace::ChallengeExpired);
+            return Err(GwMonorailError::UnlockAuthFailed);
+        }
+
+        // Check that the response is valid for our challenge.  Right now,
+        // there's only one option, but we'll use a match to check for the
+        // possibility of new types being added.
+        match (challenge, response) {
+            (
+                UnlockChallenge::EcdsaSha2Nistp256(data),
+                UnlockResponse::EcdsaSha2Nistp256 { key, signature },
+            ) => verify_signature(&data, &key, &signature),
+        }?;
+        ringbuf_entry!(Trace::UnlockAuthSucceeded);
+
+        // Pick how long we'll trust things for
+        let now = sys_get_timer().now;
+        let trust_until = now + u64::from(time_sec) * 1000;
+
+        // Reconfigure the management network for arbitrary SP access
+        if let Err(e) = self.monorail.unlock_vlans(trust_until) {
+            ringbuf_entry!(Trace::MonorailUnlockFailed(e));
+            return Err(GwMonorailError::UnlockFailed);
+        }
+
+        // Reconfigure the net task to temporarily trust this port's VLAN
+        let net = task_net_api::Net::from(crate::NET.get_task_id());
+        if let Err(e) = net.trust_vlan(vid, trust_until) {
+            ringbuf_entry!(Trace::TrustVLanFailed(e));
+            return Err(GwMonorailError::UnlockFailed);
+        }
+        ringbuf_entry!(Trace::UnlockedUntil(vid, trust_until));
+
+        // Reconfigure our own internal state to accept messages
+        self.locked[vlan_index] = LockState::UnlockedUntil(trust_until);
+
+        Ok(())
+    }
+
+    fn lock(&mut self, vid: u16) -> Result<(), GwMonorailError> {
+        ringbuf_entry!(Trace::Locking(vid));
+
+        let net = task_net_api::Net::from(crate::NET.get_task_id());
+        if let Err(e) = net.distrust_vlan(vid) {
+            ringbuf_entry!(Trace::DistrustVLanFailed(e));
+            return Err(GwMonorailError::UnlockFailed);
+        } else if let Err(e) = self.monorail.lock_vlans() {
+            ringbuf_entry!(Trace::MonorailUnlockFailed(e));
+            return Err(GwMonorailError::UnlockFailed);
+        }
+
+        let i = VLANS.iter().position(|v| v.vid == vid).unwrap();
+        self.locked[i] = LockState::Locked;
```

Comment:
If `net.distrust_vlan(_)` succeeds but `monorail.lock_vlans()` fails, should we update our state?